lib/improver/weights.py

```diff
@@ -0,0 +1,251 @@
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module to create the weights used to Blend data."""
+
+
+import numpy as np
+import iris
+
+
+def normalise_weights(weights):
+    """Ensures all weights add up to one
+
+        Args:
+            weights : array of weights.
+
+        Returns:
+            normalised_weights : array of weights where
+                                 sum = 1.0.
+    """
+    if weights.min() < 0.0:
+        msg = 'Weights must be positive, at least one value < 0.0'
+        raise ValueError(msg)
+
+    sumval = weights.sum()
+    if sumval == 0:
+        msg = 'Sum of weights must be > 0.0'
+        raise ValueError(msg)
+
+    normalised_weights = weights / sumval
+    return normalised_weights
+
+
+def nonlinear_weights(num_of_weights, cval):
+    """Create nonlinear weights
+
+        Args:
+            num_of_weights : Positive Integer - Number of weights to create.
+
+            cval : Float -  greater than 0.0 but less than or equal to 1,0,
+                   to be used for the nonlinear weights function.
+                   1.0 = equal weights for all.
+
+                   Weights will be calculated as
+                     cval**(tval-1)/Sum(of all weights)
+                   tval is the value of 1 to num_of_weights,
+
+        Returns:
+            weights : array of weights, sum of all weights = 1.0
+
+    """
+    if not isinstance(num_of_weights, int) or num_of_weights <= 0:
+        msg = ('Number of weights must be integer > 0, '
+               'num = {0:s}'.format(str(num_of_weights)))
+        raise ValueError(msg)
+    if cval <= 0.0 or cval > 1.0:
+        msg = ('cval must be greater than 0.0 and less '
+               'than or equal to 1.0 '
+               'cval = {0:s}'.format(str(cval)))
+        raise ValueError(msg)
+
+    weights_list = []
+    for tval_minus1 in range(0, num_of_weights):
+        weights_list.append(cval**(tval_minus1))
+
+    weights = normalise_weights(np.array(weights_list))
+
+    return weights
+
+
+def linear_weights(num_of_weights, y0val=1.0, slope=0.0,
+                   ynval=None):
+    """Create linear weights
+
+        Args:
+            num_of_weights : Positive Integer: Number of weights to create.
+            y0val = positive float:
```

Comment:
I'm a big fan of your style of listing arguments elsewhere in this routine and in your blending plugin (I've copied it in my stuff), so it would be nice to see it in this little bit too. So if you could replicate your style used elsewhere here (name : type \n \t description) that would be grand.

```
    Args:                                                      
        cube : iris.cube.Cube                                  
               Cube to blend across the coord.                 
        coord : string                                         
                The name of a coordinate dimension in the cube.
````